### ISSUE TITLE
Thunderbird 68 support

### ISSUE THREAD

**Issue by tasn** [trlkly/dorando-keyconfig#30]
Please add support for Thunderbird 68. I can't use thunderbird without this extension!!! :)

Thanks!

**Comment by jikamens** [comment]
I tried and failed a few months ago. :-)

It's not impossible, it's just that I didn't have the knowledge and experience necessary to accomplish it at the time and didn't have time to learn it.

I'm hoping someone else will step up to help, but so far no one has.

In the meantime, you can use my [userChromeJS](https://addons.thunderbird.net/en-us/thunderbird/addon/userchromejs-2/) add-on to run JavaScript which modifies key bindings, though that's obviously not as user-friendly as keyconfig.

Here's some sample code from my own userChrome.js file:

```JavaScript
function userChromeUnbindKey(keyset_id, key_id) {
    var keyset = document.getElementById(keyset_id);
    if (! keyset) {
        console.log("Could not find keyset " + keyset_id);
        return;
    }
    var key = document.getElementById(key_id);
    if (! key) {
        console.log("Could not find key " + key_id);
        return;
    }
    keyset.removeChild(key);
}

function userChromeBindKey(keyset_id, key, key_id, modifiers, command, observes) {
    var keyset = document.getElementById(keyset_id);
    if (! keyset) {
        console.log("Could not find keyset " + keyset_id);
        return;
    }
    var elt = document.createXULElement("key");
    if (key)
        elt.setAttribute("key", key);
    if (key_id)
        elt.setAttribute("key_id", key);
    if (modifiers)
        elt.setAttribute("modifiers", modifiers);
    if (command)
        elt.setAttribute("oncommand", command);
    if (observes)
        elt.setAttribute("observes", observes);
    keyset.appendChild(elt);
}
```

I use this, e.g., to bind Alt-S in the main Thunderbird window to a custom function I wrote:

```JavaScript
if (location == "chrome://messenger/content/messenger.xul")
    userChromeBindKey("mailKeys", "s", "key_to_spamcop", "alt",
                      "jikMoveSpam();");
```

**Comment by tasn** [comment]
Ah, thanks! I'll try to play with it and see if I can manage. :)

**Comment by willsALMANJ** [comment]
I came here just to ask if there were plans to port keyconfig to Thunderbird 68. I tried a bit to get keyconfig working in Thunderbird 68 myself but got stuck trying to get `keyconfig-service.js` to load.

I decided it was quicker to start over from scratch and build back up, testing that each part worked rather than try to reverse engineer which parts of keyconfig no longer worked and what had to change to make them work again. The addon I ended up with is https://github.com/willsALMANJ/tbkeys. It has my preferred bindings hard-coded into it and does provide for customization, but others could take it and edit in their own bindings.

If no one figures out how to port keyconfig to Thunderbird 68, I could add a UI like keyconfig's for assigning keys to javascript snippets to my extension, so it could be used more easily by others, but for now I will wait to see anyone is able to get keyconfig itself working.

**Comment by tasn** [comment]
@willsALMANJ, thanks a lot, I also want it for the same keybindings (vim-like).

How do I build the extension? I couldn't find any resources online. Thanks!

**Comment by willsALMANJ** [comment]
@tasn Zip everything inside the addon folder (i.e. there should be no `addon/` folder inside the zip archive; `manifest.json`, etc. should be the top level) and save it as a file ending in `.xpi`. Then use "Install Add-on from File.." in the Add-ons Manager.

**Comment by tasn** [comment]
That's exactly what I did, it doesn't work. It correctly installs the addon, but doesn't actually bind the keys.
However, if I go to "Debug addons -> temporary install" and install by choosing the manifest file, it works (though only until I restart thunderbird). Weird.

**Comment by morat523035** [comment]
Here is a Dorando keyconfig version that supposedly works with Firefox 68.

http://forums.mozillazine.org/viewtopic.php?p=14845317#p14845317

**Comment by willsALMANJ** [comment]
Something to be aware of: [Thunderbird is dropping support of traditional add-ons completely](https://groups.google.com/forum/#!topic/tb-planning/-E8Yw6POxEE) with the next release. So trying to get something working beyond the userChrome.js workaround or the little extension I made might not be worth the effort.

**Comment by tasn** [comment]
:(

Your little extension will continue working though?

**Comment by willsALMANJ** [comment]
No, my extension is a legacy extension as well. We need to look into what the Thunderbird developers are willing to add to MailExtensions for a keyboard shortcut API.

**Comment by ftoledo** [comment]
i really miss this addon on thunderbird =((

**Comment by tasn** [comment]
I tried not using it for a few days, and it was hell. @willsALMANJ saved the day, but apparently only temporarily. :(

I won't just miss it. I need it!

**Comment by ftoledo** [comment]
yeah @tasn thw @willsALMANJ extension work for me.. I just need only three shorcuts that save my life ([R]reply [A] Reply to all and [F]forward

=)
i hope that dorando-keyconfig can be ported to TB 68+

Saludos!

**Comment by rubo77** [comment]
Could you please explain again for a n00b, how to make the addon work in Thunderbird 68?

**Comment by ftoledo** [comment]
> Could you please explain again for a n00b, how to make the addon work in Thunderbird 68?

You must download or clone the source code , the put in into the extension directory at you profile

**Comment by rubo77** [comment]
I tried cloning this repo, and used tar:

```
git clone https://github.com/trlkly/dorando-keyconfig
cd dorando-keyconfig/
tar cvzf ../dorando-keyconfig.xpi *

```

then install this addon as file in thunderbird 68.1.2

but still blocked there: "... not compatible"

**Comment by ftoledo** [comment]
hi @rubo77 
You must zip the "addon" directory, not from the root directory of your clone
Saludos!

**Comment by rubo77** [comment]
> @tasn Zip everything inside the addon folder (i.e. there should be no `addon/` folder inside the zip archive; `manifest.json`, etc. should be the top level) and save it as a file ending in `.xpi`. Then use "Install Add-on from File.." in the Add-ons Manager.

thats what I did

**Comment by rubo77** [comment]
can someone please create such a working xpi archive and link it here?

**Comment by wshanks** [comment]
@rubo77 As far as I know, no one has gotten this add-on to work in Thunderbird 68. There was a link above about a version that worked in Firefox 68, but it wasn't clear that that really worked. There are workarounds like the userchromJS and tbkeys options linked to above. To get keyconfig to work in Thunderbird 68 would take more development time to debug which parts of the XUL API had been broken and find workarounds (without much documentation since XUL has been deprecated by Mozilla). Also, there is not much payoff to putting time into that debugging -- Thunderbird has announced that support for XUL addons like keyconfig will not be supported in the next release.

**Comment by rossmeissl** [comment]
For folks trying to use @jikamens [userChromeJS stopgap](https://github.com/trlkly/dorando-keyconfig/issues/30#issuecomment-530343943) to _unbind_ default shortcuts (say, to free them up for the venerable [Archive This](https://github.com/adamroach/tb-archive-this)), you may be wondering where to find `key_id`.

* Find the name of the command [here](https://dxr.mozilla.org/comm-central/source/comm/mail/locales/en-US/chrome/messenger/messenger.dtd). You may find it easiest to search for `"p"`, for example, to find what's currently assigned to the `p` key on your keyboard. In this example, you'll find:

  ``` xml
  <!ENTITY prevUnreadMsgCmd.key "p">
  ```

  In this case, the name of the command is `prevUnreadMsgCmd`.

* Now find that command [here](https://dxr.mozilla.org/comm-central/source/comm/mail/base/content/mainKeySet.inc) and take note of its `id`. For example:

  ``` xml
  <key id="key_previousUnreadMsg" key="&prevUnreadMsgCmd.key;"       oncommand="goDoCommand('cmd_previousUnreadMsg')"/>
  ```

  This yields `key_previousUnreadMsg` which is the `key_id` we need for the `userChromeUnbindKey` function.

**Comment by ftoledo** [comment]
any update for the 68+ support?
i really miss this extension =(  using ctrl+L to forward emails sucks! jjaja!

**Comment by romanho** [comment]
For the unbinding issue @rossmeissl mentioned above: I had the same issue for the 'r' key which I wanted to rebind. I did it by:

```javascirpt
function RemoveBindings(kset, key) {
	var keyset = document.getElementById(kset);
	if (!keyset) {
		console.log("Could not find keyset " + keyset_id);
		return;
	}
	for (var e of keyset.children) {
		var k = e.getAttribute("key");
		if (k == key) {
			console.log("Remove binding "+e.id+" for key "+k);
			keyset.removeChild(e);
		}
	}
}
```
in my userChrome.js, and then `RemoveBindings("mailKeys", "r")` before binding 'r' with various modifiers. This simply removes all existing bindings.

**Comment by fefabfee** [comment]
I'm just an ordinary user, so my apologies if my comments on this forum are out of line.  I have a simple observation.  The growing pains Thunderbird is going through right now seem to be similar to what Firefox went through not long ago. I'm guessing the code might be similar.  There is a developer on the Firefox side that has a shortcut add-on (https://github.com/mikecrittenden/shortkeys) that works in current version of Firefox. If the code was indeed similar, I was wondering if he could be of some help.

**Comment by koolx** [comment]
I'm thinking about updating TB to the latest version. I just want to change 2 keys: R for reply and N for new mail compose. How do I make that happen in the userChromeJS add-on?

**Comment by wshanks** [comment]
@fefabfee Unfortunately, Firefox add-ons do not translate to Thunderbird very well right now. One problem is that Firefox add-ons add a lot their user facing components (like key event listeners) through content scripts that get loaded into the current web page. This works well because in Firefox the web page is almost the entire UI. In Thunderbird, most of the UI is various elements controlled by Thunderbird. I see on Thunderbird's roadmap one item is to replace current UI with a web-based UI. Perhaps when that is done add-ons will be able to inject content scripts into the main UI that can interact with the whole UI.

**Comment by sespadam** [comment]
> 
> 
> For the unbinding issue @rossmeissl mentioned above: I had the same issue for the 'r' key which I wanted to rebind. I did it by:
> 
> ```
> function RemoveBindings(kset, key) {
> 	var keyset = document.getElementById(kset);
> 	if (!keyset) {
> 		console.log("Could not find keyset " + keyset_id);
> 		return;
> 	}
> 	for (var e of keyset.children) {
> 		var k = e.getAttribute("key");
> 		if (k == key) {
> 			console.log("Remove binding "+e.id+" for key "+k);
> 			keyset.removeChild(e);
> 		}
> 	}
> }
> ```
> 
> in my userChrome.js, and then `RemoveBindings("mailKeys", "r")` before binding 'r' with various modifiers. This simply removes all existing bindings.

Just so great! Been finally able to disable some key bidings which are quite bothering for me; at least till version 78.

**Comment by michele-bs** [comment]
Plaease bring keyconfig back to life!
Thunderbird 68 users are awaiting with confidence.

10,000$ for brave programmers.

**Comment by ftoledo** [comment]
tbkeys works, do you tried ? https://github.com/willsALMANJ/tbkeys

**Comment by rubo77** [comment]
@ftoledo Thousand thanks!!! 🥇

`tbkeys` works!!!!

👍 👍 👍 👍 👍 👍 👍 👍 👍 

I added instructions how do remap the dangerous shortcuts here:

https://github.com/willsALMANJ/tbkeys/issues/15

**Comment by navennn** [comment]
Thats great!
btw.
disable all shortcuts

```
{
    "0": "window.goDoCommand('')",
    "1": "window.goDoCommand('')",
    "2": "window.goDoCommand('')",
    "3": "window.goDoCommand('')",
    "4": "window.goDoCommand('')",
    "5": "window.goDoCommand('')",
    "6": "window.goDoCommand('')",
    "7": "window.goDoCommand('')",
    "8": "window.goDoCommand('')",
    "9": "window.goDoCommand('')",
    "a": "window.goDoCommand('')",
    "b": "window.goDoCommand('')",
    "c": "window.goDoCommand('')",
    "f": "window.goDoCommand('')",
    "j": "window.goDoCommand('')",
    "k": "window.goDoCommand('')",
    "m": "window.goDoCommand('')",
    "o": "window.goDoCommand('')",
    "p": "window.goDoCommand('')",
    "r": "window.goDoCommand('')",
    "s": "window.goDoCommand('')",
    "t": "window.goDoCommand('')",
    "u": "window.goDoCommand('')",
    "w": "window.goDoCommand('')",
    "x": "window.goDoCommand('')",
    "#": "window.goDoCommand('')",
    "]": "window.goDoCommand('')",
    "[": "window.goDoCommand('')"


}
```

**Comment by michele-bs** [comment]
Thank you all guys.

You can also refer to [my related thread on Mozillazine](http://forums.mozillazine.org/viewtopic.php?p=14868330&sid=391737813499f290887d93928e91bd1f#p14868330).

**Comment by koolx** [comment]
hey appreicate the link.. looking into it

**Comment by karmaking** [comment]
FWIW:
My keyconfig killerfeature is/was pressing a single key to move a message into a specific folder, and I was able to substitute this feature with the extension "Archive this" which offers this functionality (with presets and assignable hotkeys) quite well.

Would love to switch back to dorando-keyconfig nevertheless, though! :)

**Comment by eengstrom** [comment]
@karmaking - you might like [Nostalgy++](https://addons.thunderbird.net/en-US/thunderbird/addon/nostalgy_ng/).  There was a previous version of "Nostalgy" that stopped working well prior to TB 68, but the "++" version is updated for TB 78.

**Comment by karmaking** [comment]
Hi @eengstrom - thanks a lot! I'm giving this a try, as "Archive this" also stopped working since TB 78...
